### PR TITLE
Migration to Go SDK v2

### PR DESCRIPTION
**Pull request by lamnhoan**
### Core Library Migration
- **Client interface update** (cloudwatchlogs_actions.go): Updated interface to use v2 cloudwatchLogs client and types
- **Error Handling** (errors.go): Migrated from awserr to smithy-go error handling patterns
     - Large Change: removed `cwlErr` since aws sdk v2 doesn't support recursive errors well
     - Large Change: removed `errNilResultField` since `ResultField` is no longer a pointer field
- **Polling Logic** (poller.go): Updated QueryStatus handling and error processing
- **Stream Processing** (stream.go): Adapted to v2 SDK response structures
- **Client use update** (stopper.go, starter.go): Updated to work with v2 client interface
- **Other** (mgr.go, incite.go): Minimal changes that will not affect behavior

### Test Suite Updates
- **Comprehensive Test Migration**: Updated all test files to use v2 SDK mocks and error types
- cloudwatchlogs_actions.go: updated Mock client interface
- poller_test.go: Removed tests involving `errNilResultField`
- errors_test.go, example_test.go: Converted to v2 SDK patterns
- **Error Type Conversions**: Replaced awserr.New() calls with appropriate v2 error types:
    - InvalidParameterException, LimitExceededException, ServiceUnavailableException
    - Generic API errors using smithy.GenericAPIError

#### 1. Error Handling Migration
```go
// Before (v1)
awserr.New("InvalidParameterException", "message", nil)

// After (v2)  
&types.InvalidParameterException{Message: aws.String("message")}
```

#### 2. QueryStatus Type Safety
```go
// Before (v1)
status := cloudwatchlogs.QueryStatusComplete

// After (v2)
status := string(types.QueryStatusComplete)
```
#### 3. Client Initialization
```go
// Before (v1)
client := cloudwatchlogs.New(session)

// After (v2)
client := cloudwatchlogs.NewFromConfig(cfg)
```
## 🧪 Testing
- All existing unit tests pass with v2 SDK
- Tested the example in `README.md`
- Code coverage 
    - before changes: 99.5%
    - after changes: 98.6% 

### Documentation & Dependencies
- **README.md**: Updated example code to use v2 SDK patterns with proper imports and error handling
- **go.mod**: Cleaned up dependencies, removed v1 SDK, added v2 SDK requirements

## 📦 Dependencies
- **Minimum Go Version**: 1.22+
- **AWS SDK Go v2**: 1.36.5+
- **Removed**: All AWS SDK Go v1 dependencies

### CODE REVIEW

**Review by haweston — APPROVED** [review]

**Review by vcschapp — COMMENTED** [review]
I just reviewed the PR description so far, and it all seems to make sense.

One thing I'd like to dig into a bit more is around this area:

> - **Error Handling** (errors.go): Migrated from awserr to smithy-go error handling patterns
>     - Large Change: removed cwlErr since aws sdk v2 doesn't support recursive errors well

I'll have to look at this more closely.

- It looks to me like `cwlErr` is just used for unit testing support...
- I'm trying to understand *why* the cause chain support is in `cwlErr` as it doesn't seem to be used much.
- How will we be sure we're not re-introducing [Issue #13](https://github.com/gogama/incite/issues/13)?

>    - Large Change: removed errNilResultField since ResultField is no longer a pointer field

Fantastic! That was a major pain point of AWS SDK for Go V1. 👍

File: errors.go [hunk 1]
```diff
@@ -127,35 +124,72 @@ const (
...
 		// Throttled and Throttling.
-		if strings.Contains(strings.ToLower(x.Code()), "throttl") ||
-			strings.Contains(strings.ToLower(x.Message()), "rate exceeded") {
+		if strings.Contains(strings.ToLower(code), "throttl") ||
+			strings.Contains(strings.ToLower(code), "toomanyrequests") {
 			return throttlingClass
 		}
 
-		// Recursively examine the cause error, if any.
-		return classifyError(x.OrigErr())
```
Comment (lamnhoan) [inline errors.go:158]: Yes `cwlErr` is only used for unit testing and so I removed `cwlErr` since the error handling has changed. It no longer recursively examines the cause error. It will only unwrap the error if it an `http.ResponseError`.

Comment (lamnhoan) [reply errors.go:158]: Also, I have not tested if we are re-introducing issue #13 since that particular error doesn't seem like it can be easily replicated when testing.

However, I made sure to implement similar error handling as in commit [75d9c6b](https://github.com/gogama/incite/commit/75d9c6bdb59e98cace53413675fd3612abf01b0d) such as checking status codes and checking for throttling/rate limiting.

File: errors.go [hunk 2]
```diff
@@ -127,35 +124,72 @@ const (
...
+	// Check for specific CloudWatch Logs error types
+	var limitExceeded *types.LimitExceededException
+	if errors.As(err, &limitExceeded) {
+		return limitExceededClass
+	}
+
+	var serviceUnavailable *types.ServiceUnavailableException
+	if errors.As(err, &serviceUnavailable) {
+		return temporaryClass
+	}
```
Comment (vcschapp) [inline errors.go]: These implement `smithy.APIError` and I would suggest they go in that branch of code rather than standing alone.

File: errors.go [hunk 3]
```diff
@@ -127,35 +124,72 @@ const (
...
+			return temporaryClass
+		}
+
+		if strings.Contains(strings.ToLower(code), "limitexceeded") {
+			return limitExceededClass
+		}
+
+		if strings.Contains(strings.ToLower(code), "invalidparameterexception") {
+			return permanentClass
+		}
```
Comment (vcschapp) [inline errors.go]: This seems excessive. Unless empirically justified, I think we can stick with checking the specific error types like `*types.LimitExceededException`, *etc.*

File: errors.go [hunk 4]
```diff
@@ -127,35 +124,72 @@ const (
...
 
+		if httpErr.Unwrap() != nil {
+			return classifyError(httpErr.Unwrap())
+		}
+	}
+
+	// Check for generic API errors
+	var apiErr smithy.APIError
+	if errors.As(err, &apiErr) {
+		code := apiErr.ErrorCode()
```
Comment (vcschapp) [inline errors.go]: It seems to me that this logic is a bit different from the previous one... Using `errors.As` instead of a type assertion as before means that might immediately jump down cause chain instead of examining things step by step as previously...

One way to conceptualize this is that the new logic is basically doing an "eager" deep search where for each type being checked it is searching the entire cause chain to find if any cause "is" that error, versus the old logic which was "reluctantly" exploring the chain, always preferring to check 

An alternative that would stay closer to the original logic might be:

- Check if it *is* directly an `http.ResponseError` via type assertion. Then use your logic.
- Check it is *is* directly a `smithy.APIError` via type assertion and then under this branch, also:
    - Check if it is, directly, ` *LimitExceededException`, `*ServiceUnavailableException`.
        - I would also add here `*ServiceQuotaExceededException` as an example of `limitedExceededClass`.
    - At the end of the branch, check if it implements `errors.Unwrap` and if yes, call `classifyError` on the unwrapped cause...

**Review by vcschapp — COMMENTED** [review]
Still working through this, but here are some interim things I'm know I'm going to ask for.

1. Can you please generate a code coverage report both before and after so we can see the coverage change?
2. I want to make sure the `CONTRIBUTORS` file is as accurate as possible, would you mind adding yourself in!?
3. In order to search for tricky concurrency edge cases, I usually preferred to run the unit tests with many iterations (at least 100K) on multiple platforms.
    - Would you please do this and verify they pass?
    - If you have access to multiple platforms, *e.g.* Mac OS and Linux, would you mind running for both?

File: errors.go [hunk 5]
```diff
@@ -127,35 +124,72 @@ const (
...
+	var serviceUnavailable *types.ServiceUnavailableException
+	if errors.As(err, &serviceUnavailable) {
+		return temporaryClass
+	}
+
+	// Check for OperationError
+	var opErr *smithy.OperationError
+	if errors.As(err, &opErr) {
+		unwrappedErr := opErr.Unwrap()
+		if unwrappedErr != nil {
```
Comment (vcschapp) [inline errors.go]: Checking for `nil` is unnecessary since `errors.As` will just return `false` on any `nil`...

https://go.dev/play/p/3HyHBEoGK-I

File: errors.go [hunk 6]
```diff
@@ -127,35 +124,72 @@ const (
...
+		return limitExceededClass
+	}
+
+	var serviceUnavailable *types.ServiceUnavailableException
+	if errors.As(err, &serviceUnavailable) {
+		return temporaryClass
+	}
+
+	// Check for OperationError
+	var opErr *smithy.OperationError
```
Comment (vcschapp) [inline errors.go]: Does the AWS SDK basically wrap most exceptions from the service in `smithy.OperationError`?

It [appears so, tbh](https://github.com/aws/aws-sdk-go-v2/blob/main/service/cloudwatchlogs/api_client.go#L284-L305).

In that case, this check should be at the very tippy top of the function, and it should be again a type assertion rather than an `errors.As` deep traversal, and all it needs to do is call `classifyError` on the internal error. (Which, if it is `nil`, will result in `permanentClass` anyway.)

File: errors.go [hunk 7]
```diff
@@ -127,35 +124,72 @@ const (
...
+	if errors.As(err, &opErr) {
+		unwrappedErr := opErr.Unwrap()
+		if unwrappedErr != nil {
+			var deserializeErr *smithy.DeserializationError
+			if errors.As(unwrappedErr, &deserializeErr) {
+				return temporaryClass
+			}
```
Comment (vcschapp) [inline errors.go]: Is this logic trying to replicate the Issue #13 logic?

If yes, it's a regression. The substance of issue #13 was basically that an HTTP status 503 directly from the load balancer without any AWS API headers or response body info wasn't getting treated as `temporaryClass`. Yes, it manifested as a deserialization error, but that's only because the blank body/no error header response couldn't be deserialized.

[The fix](https://github.com/gogama/incite/commit/75d9c6bdb59e98cace53413675fd3612abf01b0d) for issue #13 was just to check the HTTP response status codes as a fallback if an AWS SDK level error couldn't be found. You've already migrated this logic above, so as long as it's migrated right, the problem should be solved.

Note that trying to classify every deserialization error as temporary is totally. Like if there's some bad code shipped on the server and it starts sending back a response that that client can't deserialize, we don't want to retry that.

File: errors_test.go [hunk 8]
```diff
@@ -190,28 +183,24 @@ func TestClassifyError(t *testing.T) {
...
 // https://github.com/gogama/incite/issues/13.
 func issue13Error(requestID string, statusCode int) error {
-	r := request.Request{
-		RequestID: requestID,
-		HTTPResponse: &http.Response{
+
+	httpResponse := &smithyhttp.Response{
+		Response: &http.Response{
 			StatusCode: statusCode,
-			Body:       ioutil.NopCloser(strings.NewReader("")),
```
Comment (vcschapp) [inline errors_test.go:197]: Thx 🙏

File: go.mod [hunk 9]
```diff
@@ -1,9 +1,31 @@
-module github.com/gogama/incite
+module github.com/gogama/incite/v2
 
-go 1.14
+go 1.22
```
Comment (vcschapp) [inline go.mod:3]: Can we go any lower? I see the Go docs currently claim V2 SDK needs `go 1.21`, can we potentially go down there?

File: cloudwatchlogs_actions.go [hunk 10]
```diff
@@ -7,34 +7,36 @@ package incite
...
 // CloudWatchLogsActions provides access to the CloudWatch Logs actions
 // which QueryManager needs in order to run Insights queries using the
 // CloudWatch Logs service.
 //
-// This interface is compatible with the AWS SDK for Go (v1)'s
-// cloudwatchlogsiface.CloudWatchLogsAPI interface and *cloudwatchlogs.CloudWatchLogs
-// type, so you may use either of these AWS SDK types to provide the
+// This interface is compatible with the AWS SDK for Go (v2)'s
+// *cloudwatchlogs.Client type, so you may use this AWS SDK type to provide the
 // CloudWatch Logs act capabilities.
```
Comment (vcschapp) [inline cloudwatchlogs_actions.go]: nit: (This typo was there before, but) we should change "act" to "action" on line 19

File: poller.go [hunk 11]
```diff
@@ -69,17 +72,17 @@ func (p *poller) manipulate(c *chunk) outcome {
 		}
 	}
 
-	if output.Status == nil {
+	if len(output.Status) == 0 {
 		c.err = &UnexpectedQueryError{c.queryID, c.stream.Text, errNilStatus()}
 		return finished
 	}
```
Comment (vcschapp) [inline poller.go]: This block doesn't seem needed anymore.

It was there before as purely defensive programming because SDK 1 over-used pointers everywhere, and dereferencing a `nil` pointer causes a panic.

Checking for empty string is not necessary since it'll hit the `default` block of the switch.

Note that empty string would have hit the `default` block *before* anyway, so this change apart from being redundant is also subtly changing the logic!

File: poller.go [hunk 12]
```diff
@@ -69,17 +72,17 @@ func (p *poller) manipulate(c *chunk) outcome {
...
 
-	status := *output.Status
+	status := output.Status
 	switch status {
-	case cloudwatchlogs.QueryStatusScheduled, "Unknown":
+	case types.QueryStatusScheduled, "Unknown":
```
Comment (vcschapp) [inline poller.go]: Looks like they have `types.QueryStatusUnknown` now (which they didn't have previously in the V1 SDK version I was working against). We should use their symbolic constant in preference to the literal value `"Unknown"`.

https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.51.0/types#QueryStatus

File: poller.go [hunk 13]
```diff
@@ -111,7 +114,7 @@ func (p *poller) manipulate(c *chunk) outcome {
 		fallthrough
 	default:
 		translateStats(output.Statistics, &c.Stats)
-		c.err = &TerminalQueryStatusError{c.queryID, status, c.stream.Text}
+		c.err = &TerminalQueryStatusError{c.queryID, string(status), c.stream.Text}
```
Comment (vcschapp) [inline poller.go:112]: Should `TerminalQueryStatusError.Status` just have the same type as `GetQueryResultsOutput.Status`?

File: starter.go [hunk 14]
```diff
@@ -48,15 +48,20 @@ func (s *starter) manipulate(c *chunk) outcome {
 	starts := epochMillisecond(c.start)
 	ends := epochMillisecond(c.end.Add(-time.Millisecond)) // CWL uses inclusive time ranges, we use exclusive ranges.
 
+	// Convert Limit from int64 to int32 for AWS SDK Go v2
+	limit := int32(c.stream.Limit)
```
Comment (vcschapp) [inline starter.go]: Rather than doing this, it feels safer to change the data type in the field `QuerySpec.Limit` from `int64` to `int32`.

A straight cast here is risky, because it's changing the data type without accounting for possible overflow.

https://go.dev/play/p/yEfg_AY270T

Yes, technically, we check it against `MaxLimit` but the best way to do things is to maintain the same type all the way through the API to reduce the possibility of error.

FWIW, `int64` was chosen previously because that was the value accepted by AWS SDK for Go V1.

**Review by vcschapp — COMMENTED** [review]
I'm actually pleasantly surprised by how non-invasive the change was!

Nice work :)

Left some comments. Once it's all addressed, all I want to do is triple check the `classifyError` unit test logic and then I should think we're GTG here.

File: go.mod (hunk 9 above)
Comment (lamnhoan) [reply go.mod:3]: Interesting. Where are you seeing that?
On the github for aws-sdk-go-v2, I see `1.22`. 
https://github.com/aws/aws-sdk-go-v2#:~:text=The%20v2%20SDK%20requires%20a%20minimum%20version%20of%20Go%201.22.

File: cloudwatchlogs_actions.go (hunk 10 above)
Comment (lamnhoan) [reply cloudwatchlogs_actions.go]: Ok sounds good.

File: errors.go (hunk 4 above)
Comment (lamnhoan) [reply errors.go]: Oh I see. This was very insightful. I will make these changes. Also, `smithy.APIError` doesn't implement `errors.Unwrap` so checking this will not be necessary. I will check for `errors.Unwrap` right before the end of the function.

File: errors.go (hunk 2 above)
Comment (lamnhoan) [reply errors.go]: I will reply to the comment above this ^ one here.
1. Code coverage report has been added to the PR description.
2. Yes I can add myself in
3. Ok I will run the unit tests with many iterations on Mac OS and Linux.

Comment (lamnhoan) [reply errors.go]: Sounds good. I will make these changes.

File: errors.go (hunk 5 above)
Comment (lamnhoan) [reply errors.go]: Ah I see. Thanks for letting me know. Will make the change.

File: errors.go (hunk 6 above)
Comment (lamnhoan) [reply errors.go]: Noted.

File: errors.go (hunk 7 above)
Comment (lamnhoan) [reply errors.go]: Yes it was trying to replicate the logic. Ah I see now. Thanks for the insight!

File: poller.go (hunk 11 above)
Comment (lamnhoan) [reply poller.go]: Oh nice catch. Will make change

File: poller.go (hunk 12 above)
Comment (lamnhoan) [reply poller.go]: Nice catch. I will make this change.

File: poller.go (hunk 13 above)
Comment (lamnhoan) [reply poller.go:112]: The `Status` field in `TerminalQueryStatusError` was kept as `string` instead of `QueryStatus` to allow testing with invalid status values that aren't part of the QueryStatus enumeration.
This is seen in some unit tests.

An example is here: https://github.com/lamnhoan/incite/blob/6e7b09605ce28e1de09fe065de56d1ce0ee79b18/incite_test.go#L355

File: starter.go (hunk 14 above)
Comment (lamnhoan) [reply starter.go]: Noted. Will make change.

File: errors.go (hunk 3 above)
Comment (lamnhoan) [reply errors.go]: Yes I will make this change. Originally had this here to migrate over unit tests that had generic `awserr` errors that didn't use any of the `cloudwatchlogs` error codes. I will also make changes to those unit tests.

File: go.mod (hunk 9 above)
Comment (vcschapp) [reply go.mod:3]: I'm not sure where I saw this now - and it is possible that I was looking at a stale document.

Agree that your change seems correct. Thanks for the reference link!

File: poller.go (hunk 13 above)
Comment (vcschapp) [reply poller.go:112]: NB: The unit tests can technically cast any string to that type, *e.g.* `QueryStatus("some invalid value")`. But I'm good with leaving it how it is.

File: errors.go [hunk 15]
```diff
@@ -127,35 +119,46 @@ const (
...
-		// Short-circuit if the HTTP status code indicates retryability.
-		if f, ok := err.(awserr.RequestFailure); ok {
-			status := f.StatusCode()
-			if status == 429 {
-				return throttlingClass
-			} else if status == 502 || status == 503 || status == 504 {
-				return temporaryClass
-			}
+
+	// Check for OperationError
```
Comment (vcschapp) [inline errors.go]: nit: I would delete this comment. Doesn't add much to the actual code.

File: errors.go [hunk 16]
```diff
@@ -127,35 +119,46 @@ const (
...
 		}
+	}
 
-		// Check for known CloudWatch Logs retryability codes.
-		switch x.Code() {
-		case cloudwatchlogs.ErrCodeLimitExceededException:
-			return limitExceededClass
-		case cloudwatchlogs.ErrCodeServiceUnavailableException:
+	// Check for ResponseError
+	// Check HTTP status codes
```
Comment (vcschapp) [inline errors.go]: nit: I would delete these comment lines. They don't seem to add much to the code.

File: errors.go [hunk 17]
```diff
@@ -127,35 +119,46 @@ const (
...
+		case *types.LimitExceededException,
+			*types.ServiceQuotaExceededException:
+			return limitExceededClass
+		case *types.InvalidOperationException,
+			*types.InvalidParameterException,
+			*types.MalformedQueryException,
+			*types.ResourceNotFoundException,
+			*types.UnrecognizedClientException:
+			return permanentClass
+		}
```
Comment (vcschapp) [inline errors.go]: Why not just use `default` here?

File: errors.go [hunk 18]
```diff
@@ -127,35 +119,46 @@ const (
...
+		case 502, 503, 504:
 			return temporaryClass
 		}
 
-		// Check for throttling using common AWS service patterns for indicating
-		// throttling via exception. Omit 'e' suffix on 'throttl' to match
-		// Throttled and Throttling.
-		if strings.Contains(strings.ToLower(x.Code()), "throttl") ||
-			strings.Contains(strings.ToLower(x.Message()), "rate exceeded") {
-			return throttlingClass
```
Comment (vcschapp) [inline errors.go]: I would ideally like to include this logic at the bottom of the `Smithy.APIError` block.

(Where `x.Code()` becomes `x.ErrorCode()` and `x.Message()` becomes `x.ErrorMessage()`.)

File: errors.go [hunk 19]
```diff
@@ -127,35 +119,46 @@ const (
...
-	if x, ok := err.(awserr.Error); ok {
-		// Short-circuit if the HTTP status code indicates retryability.
-		if f, ok := err.(awserr.RequestFailure); ok {
-			status := f.StatusCode()
-			if status == 429 {
-				return throttlingClass
-			} else if status == 502 || status == 503 || status == 504 {
-				return temporaryClass
-			}
+
```
Comment (vcschapp) [inline errors.go]: nit: extra newline here not needed between start of function and first line of code

File: errors_test.go [hunk 20]
```diff
@@ -190,28 +174,24 @@ func TestClassifyError(t *testing.T) {
 // issue13Error returns an error of the type that triggered issue #13,
 // https://github.com/gogama/incite/issues/13.
 func issue13Error(requestID string, statusCode int) error {
-	r := request.Request{
-		RequestID: requestID,
-		HTTPResponse: &http.Response{
+
```
Comment (vcschapp) [inline errors_test.go]: nit: Again, unnecessary newline between function prototype and first line of code.

File: errors_test.go [hunk 21]
```diff
@@ -190,28 +174,24 @@ func TestClassifyError(t *testing.T) {
...
+	httpResponse := &smithyhttp.Response{
+		Response: &http.Response{
```
Comment (vcschapp) [inline errors_test.go:179]: To get maximum parity with the previous code, suggest you use the [AWS SDK's version for this error](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2@v1.36.5/aws/transport/http#ResponseError), which has a `RequestID` field built in.

Unfortunately it looks like another case we need to handle explicitly in `classifyError`...

File: errors.go (hunk 18 above)
Comment (vcschapp) [inline errors.go]: I would like to retain this logic.

File: errors.go [hunk 22]
```diff
@@ -127,35 +119,46 @@ const (
...
-		case cloudwatchlogs.ErrCodeServiceUnavailableException:
+	// Check for ResponseError
+	// Check HTTP status codes
+	if httpErr, ok := err.(*http.ResponseError); ok {
+		switch httpErr.HTTPStatusCode() {
+		case 429:
+			return throttlingClass
+		case 502, 503, 504:
 			return temporaryClass
 		}
```
Comment (vcschapp) [inline errors.go]: Unfortunately, I just realized that AWS SDK *and* Smithy *both* have their own versions of `http.ResponseError` and they aren't interchangeable. The AWS one is composed of the Smithy one in a weird way.

I think ultimately what we need is something along the lines of:

```go
if httpErr, ok := err.(*awshttp.ResponseError); ok {
    return classifyError(httpErr.ResponseError)
}

if httpErr, ok := err.(*smithyhttp.ResponseError); ok {
    // Main handling that checks the status code, but also retains the `strings.Contains` logic...
}

**Review by vcschapp — COMMENTED** [review]
Just about there. A few more comments on the line.

File: errors.go (hunk 15 above)
Comment (lamnhoan) [reply errors.go]: Noted.

File: errors.go (hunk 16 above)
Comment (lamnhoan) [reply errors.go]: Noted.

File: errors.go (hunk 17 above)
Comment (lamnhoan) [reply errors.go]: Good point. I didn't put a `default` initially since I had thought other `Exception` types would be caught by the blocks below. But it seems like all other `Exception` types would hit `permanentClass` at the end of the function.

File: errors.go (hunk 18 above)
Comment (lamnhoan) [reply errors.go]: I will put this at the top of the `Smithy.APIError` block since this won't reach the bottom if we have a `default` statement

File: errors.go (hunk 19 above)
Comment (lamnhoan) [reply errors.go]: Noted.

File: errors_test.go (hunk 20 above)
Comment (lamnhoan) [reply errors_test.go]: Noted.

File: errors.go (hunk 18 above)
Comment (lamnhoan) [reply errors.go]: Noted.

File: errors.go (hunk 22 above)
Comment (lamnhoan) [reply errors.go]: Gotcha.

File: errors_test.go (hunk 21 above)
Comment (lamnhoan) [reply errors_test.go:179]: Noted.

File: errors.go (hunk 22 above)
Comment (lamnhoan) [reply errors.go]: `smithyhttp.ResponseError` doesn't have `ErrorCode()` or `Message()`. Should this just go in the `smithy.APIError` block?

Comment (vcschapp) [reply errors.go]: My interpretation is they have to go at the top-level of the `classifyError` function, so you end up with type assertion `if` statements one after another for:

1. `*smithyhttp.OperationError`
2. `*awshttp.ResponseError`
3. `*smithyhttp.ResponseError`
4. `smithy.APIError`

What do you think?

**Review by vcschapp — APPROVED** [review]
Thanks Lam, the changes look good to me!

Ship it!

```
       :::::::               _H_         _,;8ba    ""                '
  ':::'       ::::          '   ;   ::  | :8888P    ""             "
                         ,==O| |        | ,::""'            ::::::  '
 :::::'           ::::  ;#$" | |    v   | |
        /"V"\           :#   | |        | |              \   /
                         "*-O| |  :::  O| |               " "
                            /| |      / | |
            ':::::::       ;'| |     ;: | |              :::::
                 ::::'    A  | |     |: | |                ""  ::::
          ':::::         /;i | |    ;'  | |      /" "\         ""
    /"-"\         "     ;  | | |   j  ; | |
                ""     ;;  | | |  /  :: | |              ,,
                      ,":  | | | ,  ;:  | |           '::::::::::,
         ~    "      /; :: | | |j     , | |            :::::   ""
  ':::'     ::::   ,d"  '::A | |: ;  :: | |   v          ""::::~
          :::::'   :  ,:: ::H| || ::::,:| |            "::::
   '::::'          : ;::  ::H| || :: ,: | |                 ::::
             v     |  :: ;: H| ||  : :: | | ''              ""
                   | ;:   :: | |!    :  | |
                   |         | | \  \ \ | |
                    \  \  \  | |  \ \  \| |        i-i-i-i-i-i
            "::::    \  \  \ | |   O O O| |       ,###########b
    ####b             O  O  O| |        | |      ,###HH##HH####
    $####b ,I.I.I.I.   ===== | |========| |==  ,#####HH##HH#####
     $$##############  ====================== ###############II#
        "# ######################################WWWWWWWWWW##II#
         "#WWWWWWWWWW_WWWWWWW_WWWWWWW_WWWWWWWW_WWWWWW_WWWW#####
          "#########(0)#####(0)#####(0)######(0)####(0)######"
             "##WWWWW"WWWWWWW"WWWWWWW"WWWWWWWW"WWWWWW"WW##"
  , ; ,;:,      #################################'''####" JUANSRX:
;" " " #~~;""''  "#####;~~~~ ##  ,#''####;;~ ~##;;#;;''
 ""--~~"",,--   ""  :::~~=~:::::::::;;:----~~~""--:::~;;::  ;;;  ;;;
```
